sympy/core/operations.py

```diff
@@ -490,3 +491,72 @@ def args(self):
     @staticmethod
     def _compare_pretty(a, b):
         return (str(a) > str(b)) - (str(a) < str(b))
+
+class AssocOpDispatcher(Dispatcher):
+    """
+    Dispatcher for associative operators
+
+    Explanation
+    ===========
+
+    This class lets the user to add or multiply different types of
+    arguments. It is recommended to define the combination by registering
+    with this class rather than complicating `Add.flatten` and `Mul.flatten`.
+
+    Examples
+    ========
+
+    >>> from sympy import Mul, Expr, Symbol
+    >>> from sympy.vector import CoordSys3D, Vector, VectorMul
+    >>> from sympy.core.operations import AssocOpDispatcher
+    >>> a, i = Symbol('a'), CoordSys3D('C').base_vectors()[0]
+
+    >>> mul = AssocOpDispatcher('mul', Mul)
+
+    >>> @mul.register(Expr, Vector)
+    ... def _(a,b, **kwargs):
+    ...     return VectorMul(a, b, **kwargs)
+
+    >>> mul()
+    1
+    >>> mul(a)
+    a
+    >>> mul(i)
+    C.i
+
+    >>> mul(a, i)
+    a*C.i
+    >>> mul(i, a)
+    a*C.i
+    >>> mul(a, i, a)
+    a**2*C.i
+
+    """
+    def __init__(self, name, fallback, doc=None):
+        self.fallback = fallback
+        super().__init__(name, doc)
+
+    def __call__(self, *args, **kwargs):
+
+        if len(args) < 2:
+            try:
+                return super().__call__(*args, **kwargs)
+            except NotImplementedError:
+                return self.fallback(*args, **kwargs)
+
+        elif len(args) == 2:
+            try:
+                try:
+                    return super().__call__(*args, **kwargs)
+                except NotImplementedError:
+                    return super().__call__(*args[::-1], **kwargs)
+            except NotImplementedError:
+                return self.fallback(*args, **kwargs)
```

Comment:
You've assumed commutativity here, but all you're allowed to assume is associativity.

Reply:
Oops. You're right. I'll fix this.